src/descriptor/create_descriptor.rs

```diff
@@ -0,0 +1,468 @@
+//! Tools for creating Descriptor and witness stack from given scriptpubkey and corresponding
+//! scriptsig and witness.
+//!
+
+use bitcoin::{self, Script, PublicKey};
+
+use descriptor::Descriptor;
+use miniscript::Miniscript;
+use descriptor::satisfied_contraints::StackElement;
+use Error;
+use ToPublicKey;
+use bitcoin::blockdata::script::Instruction;
+use bitcoin::blockdata::opcodes;
+use hash160;
+
+/// Helper function for creating StackElement from Push instructions. Special case required for
+/// handling OP_PUSHNUM_1.
+/// Dissatisfied is Pushbytes0 and witness are mapped to Pushbytes instruction in scriptsig and
+/// satisfied is mapped as OP_PUSHNUM_1.
+/// Other opcodes are considered Non-standard in bitcoin core.
+/// Miniscript should not use other pushes apart from PUSHNUM_1. This will err on reciving anything
+/// which is not PUSHBYTES, OR PUSHNUM_1 as other things are expected to happen in Miniscript
+///
+/// NOTE: Miniscript pushes should only be either boolean, 1 or 0, signatures, and hash preimages.
+/// As per the current implementation, PUSH_NUM2 results in an error
+fn instr_to_stackelem<'txin>(ins: &Instruction<'txin>) -> Result< StackElement<'txin>, Error> {
+    match *ins {
+        //Also covers the dissatisfied case as PushBytes0
+        Instruction::PushBytes(v) => Ok(StackElement::from(v)),
+        Instruction::Op(opcodes::all::OP_PUSHNUM_1) => Ok(StackElement::Satisfied),
+        _ => Err(Error::BadDescriptor),
+    }
+}
+
+/// Helper function which splits the scriptsig into 2 parts returns the corresponding elements.
+/// Usually used for scripts which have top element as Pk (p2pkh) or redeem script(p2sh).
+/// Converts the other script elements into Vec<StackElement>
+fn parse_scriptsig_top(script_sig: &bitcoin::Script) -> Result<(Vec<u8>, Vec<StackElement>), Error>
+{
+    let instructions : Vec<Instruction> = script_sig.iter(true).collect();
```

Comment:
You could do the stack element conversion here and avoid an allocation:
```
let stack: Vec<StackElement> = script_sig
    .iter(true)
    .map(instr_to_stackelem)
    .collect()?
```
Then you can pop the last element off of the stack like
```
if let Some(StackElement::Push(pk_bytes)) = stack.pop() {
    Ok(pk_bytes.to_vec(), stack)
} else {
    Err(Error::BadDescriptor)
}
```
I'm also not thrilled about the `.to_vec` in there, but it seems necessary to use `Script::to_p2sh` on the result, so I'll allow it. (The alternative is to do the `to_p2sh` conversion manually.)